core/query/src/main/java/org/eclipse/rdf4j/query/QueryResults.java

```diff
@@ -509,20 +500,27 @@ private static boolean bindingSetsMatch(BindingSet bs1, BindingSet bs2, Map<BNod
 	 * @return true if compatible
 	 */
 	public static boolean bindingSetsCompatible(BindingSet bs1, BindingSet bs2) {
-		Set<String> sharedBindings = new HashSet<>(bs1.getBindingNames());
-		sharedBindings.retainAll(bs2.getBindingNames());
+		Set<String> bs1BindingNames = bs1.getBindingNames();
+		if (bs1BindingNames.isEmpty())
+			return true;
 
-		for (String bindingName : sharedBindings) {
-			Value value1 = bs1.getValue(bindingName);
-			Value value2 = bs2.getValue(bindingName);
+		Set<String> bs2BindingNames = bs2.getBindingNames();
 
-			if (value1 == null || value2 == null) {
-				// variable is unbound in one set, therefore compatible.
-				continue;
-			}
+		for (String bindingName : bs1BindingNames) {
+
+			if (bs2BindingNames.contains(bindingName)) {
+				Value value1 = bs1.getValue(bindingName);
+
+				// if a variable is unbound in one set it is compatible
+				if (value1 != null) {
+					Value value2 = bs2.getValue(bindingName);
+
+					// if a variable is unbound in one set it is compatible
+					if (value2 != null && !value1.equals(value2)) {
+						return false;
+					}
+				}
 
-			if (!value1.equals(value2)) {
-				return false;
 			}
 		}
```

Comment:
This method gets called a lot for queries with `MINUS`. Instantiating a new HashMap every time is very expensive, so I've rewritten it so we don't need to.